packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts

```diff
@@ -242,20 +211,16 @@ export default createRule<Options, MessageIds>({
 
         const castType = services.getTypeAtLocation(node);
 
+        const grandparent = node.parent.parent!;
+
         if (
-          isTypeFlagSet(castType, ts.TypeFlags.Literal) ||
-          (tsutils.isObjectType(castType) &&
-            (tsutils.isObjectFlagSet(castType, ts.ObjectFlags.Tuple) ||
-              couldBeTupleType(castType)))
+          // It's not safe to remove a cast to a literal type, unless we are in a `const` variable declaration, as that
+          // type would otherwise be widened without the cast.
+          ((grandparent.type === AST_NODE_TYPES.VariableDeclaration &&
+            grandparent.kind === 'const') ||
+            !castType.isLiteral()) &&
+          services.getTypeAtLocation(node.expression) === castType
         ) {
-          // It's not always safe to remove a cast to a literal type or tuple
-          // type, as those types are sometimes widened without the cast.
-          return;
```

Comment:
[style] IMO having two ifs is much clearer than one.
Personally I find that it can be much harder to read a bunch of mixed logic that's also of mixed negation.